src/erl/typespec.rs

```diff
@@ -0,0 +1,263 @@
+    let return_type_params = extract_params_from_type(return_type);
+    "-spec "
+        .to_doc()
+        .append(atom(name.to_string()))
+        .append(external_spec_args(args, &return_type_params))
+        .append(" -> ")
+        .append(typ_to_erl(return_type))
+        .append(".")
+}
+
+fn external_spec_args(args: &[ExternalFnArg], allowed_params: &[String]) -> Document {
+    wrap_args(
+        args.iter()
+            .map(|arg| type_ast_name(&arg.typ, Some(allowed_params))),
+    )
+}
+
+// Extracts the type names from a TypeAst - We optionally pass in a list of type parameters that we
+// are aware of.
+fn type_ast_name(ast: &TypeAst, allowed_params: Option<&[String]>) -> Document {
+    match ast {
+        TypeAst::Var { name, .. } =>
+        // If we provide a list of allowed type parameter names then we're in a context with
+        // limited possible type parameter names and if we come across one that isn't 'allowed'
+        // then we're better just outputting 'any' instead. This prevents unknown or unused
+        // type parameter warnings from erlang
+        {
+            allowed_params
+                .map(|slice| {
+                    if slice.contains(name) {
+                        name.to_camel_case()
+                    } else {
+                        "any".to_string()
+                    }
+                })
+                .unwrap_or(name.to_camel_case())
+                .to_doc()
+        }
+        TypeAst::Constructor { name, args, .. } => type_to_erl(name, &args, allowed_params),
+        _ => "term".to_doc(),
+    }
+}
+
+fn type_to_erl(name: &str, args: &[TypeAst], allowed_params: Option<&[String]>) -> Document {
+    match name {
+        "String" => "binary()".to_doc(),
+        "Int" => "integer()".to_doc(),
+        "Bool" => "boolean()".to_doc(),
+        "List" => "list".to_doc().append(wrap_args(
+            args.iter().map(|arg| type_ast_name(arg, allowed_params)),
+        )),
+        "Map" => concat(
+            args.iter()
+                .map(|arg| type_ast_name(arg, allowed_params))
+                .intersperse(" => ".to_doc()),
+        )
+        .surround("#{", "}"),
+        _ => name.to_snake_case().to_doc(),
+    }
+}
+
+/* Extraction functions - There are various places where we want to make sure that we're only
+   outputting type parameters if we know they are going to be used in order to avoid unused type
+   parameter warnings. Consequently we have these functions to help us walk the type trees and find
+   what we need.
+*/
+fn extract_params_from_constructors(constructors: &[RecordConstructor]) -> Vec<String> {
+    let mut set = std::collections::HashSet::new();
+    constructors
+        .iter()
+        .flat_map(|constructor| {
+            constructor
+                .args
+                .iter()
+                .flat_map(|(_, ast, ..)| extract_params_from_ast(&ast))
+                .collect::<Vec<String>>()
```

Comment:
Would be nice to allocate fewer vectors. Perhaps we could pass a mutable reference to a vector into this function.

Reply:
Certainly understand this is less than desirable. Happy to come back to it if we get the branch into an otherwise functioning state.